mount/mount_linux.go

```diff
@@ -51,45 +91,81 @@ func (m *Mount) mount(target string) (err error) {
 		}
 	}
 	var (
-		chdir   string
-		options = m.Options
+		chdir     string
+		recalcOpt bool
+		usernsFd  int
+		options   = m.Options
 	)
+	opt := parseMountOptions(options)
+	// The only remapping of both GID and UID is supported
+	if opt.uidmap != "" && opt.gidmap != "" {
+		var (
+			childProcCleanUp func()
+		)
+		if usernsFd, childProcCleanUp, err = GetUsernsFD(opt.uidmap, opt.gidmap); err != nil {
+			return err
+		}
+		defer childProcCleanUp()
+
+		// overlay expects lowerdir's to be remapped instead
+		if m.Type == "overlay" {
+			var (
+				userNsCleanUp func()
+			)
+			options, userNsCleanUp, err = prepareIDMappedOverlay(usernsFd, options)
+			defer userNsCleanUp()
```

Comment:
This cleanup here is always called when the function end (the mount function). So, we will try to umount the lower layers always after a mount. Is this on purpose? We don't do this with non-idmap lower layer, right? Is there a reason to do this or not?

Reply:
You mean why we don't do this for non-idmapped overlay? Could, you, please, clarify?

P.S.: if mount options have `uidmap != ""` and `gidmap != ""` then we check if this is an overlay mount by `if m.Type == "overlay"`. This way we're sure that we should use idmapped mounts for overlay, so once we created those mounts we should clean them up. If we're not going to use idmapped mounts for overlay, options are not going to be changed, so no need to call this "cleanup".

You can check that by trying `$ mount` there will be different lowedirs (which are actually removed once we exited `mount(...)`).

Reply:
No, I'm not talking about non-idmap layers. What I want to say is: you defer userNsCleanup(), that means that when this function ends, that will be called.

In this case, this is done inside the "mount()" function, so when the mount function ends, it will call that. That means it will unmount the lower layers and delete the files.

Do we want to do that only if we find errors or do we want to do it unconditionally (as we are doing now)?

Reply:
It's always better to remove those directories anyway, podman also works this way

Reply:
> In this case, this is done inside the "mount()" function, so when the mount function ends, it will call that. That means it will unmount the lower layers and delete the files.

It's safe to umount the temp mountpoint and delete it.

Reply:
Perfect, thanks!